src/charm.py

```diff
@@ -191,19 +190,12 @@ def _on_juju_info_joined(self, _event):
 
     def _on_cos_data_changed(self, event):
         """Trigger renewals of all data if there is a change."""
-        try:
-            self._connect_logging_snap_endpoints()
-            self._update_config()
-            self._update_status()
-            self._update_metrics_alerts()
-            self._update_loki_alerts()
-            self._update_grafana_dashboards()
-        except MultiplePrincipalsError:
-            logger.error(
-                "Multiple applications claiming to be principle. Update the cos-agent library in the client application charms."
-            )
-            self.unit.status = BlockedStatus("Multiple Principal Applications")
-            event.defer()
+        self._connect_logging_snap_endpoints()
```

Comment:
why is MultiplePrincipalsError no longer a valid edge case?

Reply:
Since we no longer need to figure out which charm is principal, we no longer identify principal vs subordinate. Thus multiple principal units is completely fine.